korok/korok.py

```diff
@@ -51,7 +81,89 @@ def fit(
+            global bm25s
+            import bm25s
+
+            bm25 = bm25s.BM25()
+            tokens = bm25s.tokenize(texts, stopwords=stopwords)
+            bm25.index(tokens)
+        else:
+            bm25 = None
+
+        return cls(encoder=encoder, vicinity=vicinity, reranker=reranker, bm25=bm25, alpha=alpha, corpus=texts)
+
+    def _normalize_scores(self, scores: np.ndarray) -> np.ndarray:
+        """Normalize the scores to be between 0 and 1."""
+        # Get the min and max scores for each document
+        min_score = np.min(scores, axis=-1).reshape(-1, 1)
+        max_score = np.max(scores, axis=-1).reshape(-1, 1)
+
+        # Normalize the scores
+        denom = max_score - min_score
+        numerator = scores - min_score
+
+        return numerator / denom
+
+    def _split_vicinity_results(self, results: List[List[Tuple[str, float]]]) -> List[List[Tuple[str, float]]]:
+        """Split the results from vector search into two lists."""
+        vicinity_docs = []
+        vicinity_scores = []
+        for result in results:
+            for doc, score in result:
+                vicinity_docs.append(doc)
+                vicinity_scores.append(score)
+        vicinity_scores = np.array(vicinity_scores)
+        return vicinity_docs, vicinity_scores
+
+    def _merge_results(
+        self,
+        vicinity_results: List[List[Tuple[str, np.float64]]],
+        bm25_results: Tuple[List[List[str]], np.ndarray],
+        k: int = 10,
+    ) -> List[List[tuple[str, float]]]:
+        """Merge the results from vector search and bm25 search."""
+        # Initialize the scores list
+        scores_list = [{} for i in range(len(vicinity_results))]
+
+        # Split the results into docs and scores
+        vicinity_docs, vicinity_scores = self._split_vicinity_results(vicinity_results)
+        bm25_docs, bm25_scores = bm25_results
+
+        # convert bm25 docs to strings
+        bm25_docs = [[str(doc) for doc in doclist] for doclist in bm25_docs]
+
+        # Normalize the scores
+        vicinity_scores = self._normalize_scores(vicinity_scores)
+        bm25_scores = self._normalize_scores(bm25_scores)
+
+        # Combine the docs and scores into a dictionary (to account for mismatched docs)
+        for i in range(len(vicinity_results)):
+            for doc, score in zip(vicinity_docs[i], vicinity_scores[i]):
+                scores_list[i][doc] = Document(text=doc, vicinity_score=score)
+
+            for doc, score in zip(bm25_docs[i], bm25_scores[i]):
+                if doc not in scores_list[i]:
+                    scores_list[i][doc] = Document(text=doc, bm25_score=score)
+                else:
+                    scores_list[i][doc].bm25_score = score
+
+        # Combine the docs and scores into a list of tuples
+        results = []
+        for i in range(len(vicinity_results)):
+            # Get the combined scores
+            results.append([(key, value.combine_scores(self.alpha))
```

Comment:
This append is unnecessary. You append the result here, and then immediately index below. Here's something that could also work: e.g.:

```python
result = [(key, value.combine_scores(self.alpha)) ...]
result = sorted(result, key=lambda x: x[1], reverse=True)
results[i] = result[:k]
```

Reply:
Hey @stephantul, 

In the code, since `results[i]` has not been assigned earlier for that index, assigning it directly results in an error.

```
```